cli/src/main/java/com/diffplug/spotless/cli/SpotlessCLI.java

```diff
@@ -0,0 +1,237 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.cli;
+
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.util.List;
+
+import javax.annotation.Nonnull;
+
+import com.diffplug.spotless.Formatter;
+import com.diffplug.spotless.FormatterStep;
+import com.diffplug.spotless.LineEnding;
+import com.diffplug.spotless.LintState;
+import com.diffplug.spotless.ThrowingEx;
+import com.diffplug.spotless.cli.core.FileResolver;
+import com.diffplug.spotless.cli.core.SpotlessActionContext;
+import com.diffplug.spotless.cli.core.SpotlessCommandLineStream;
+import com.diffplug.spotless.cli.core.TargetFileTypeInferer;
+import com.diffplug.spotless.cli.core.TargetResolver;
+import com.diffplug.spotless.cli.execution.SpotlessExecutionStrategy;
+import com.diffplug.spotless.cli.help.OptionConstants;
+import com.diffplug.spotless.cli.steps.GoogleJavaFormat;
+import com.diffplug.spotless.cli.steps.LicenseHeader;
+import com.diffplug.spotless.cli.steps.Prettier;
+import com.diffplug.spotless.cli.version.SpotlessCLIVersionProvider;
+
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+
+@Command(name = "spotless", mixinStandardHelpOptions = true, versionProvider = SpotlessCLIVersionProvider.class, description = "Runs spotless", synopsisSubcommandLabel = "[FORMATTING_STEPS]", commandListHeading = "%nAvailable formatting steps:%n", subcommandsRepeatable = true, subcommands = {
+		LicenseHeader.class,
+		GoogleJavaFormat.class,
+		Prettier.class})
+public class SpotlessCLI implements SpotlessAction, SpotlessCommand, SpotlessActionContextProvider {
+
+	@CommandLine.Spec
+	CommandLine.Model.CommandSpec spec; // injected by picocli
+
+	@CommandLine.Option(names = {"--mode", "-m"}, defaultValue = "APPLY", description = "The mode to run spotless in." + OptionConstants.VALID_AND_DEFAULT_VALUES_SUFFIX)
+	SpotlessMode spotlessMode;
+
+	@CommandLine.Option(names = {"--basedir"}, hidden = true, description = "The base directory to run spotless in. Intended for testing purposes only.")
+	Path baseDir;
+
+	@CommandLine.Option(names = {"--target", "-t"}, description = "The target files to format.")
+	public List<String> targets;
+
+	@CommandLine.Option(names = {"--encoding", "-e"}, defaultValue = "UTF-8", description = "The encoding of the files to format." + OptionConstants.DEFAULT_VALUE_SUFFIX)
+	public Charset encoding;
+
+	@CommandLine.Option(names = {"--line-ending", "-l"}, defaultValue = "UNIX", description = "The line ending of the files to format." + OptionConstants.VALID_AND_DEFAULT_VALUES_SUFFIX)
+	public LineEnding lineEnding;
+
+	@Override
+	public Integer executeSpotlessAction(@Nonnull List<FormatterStep> formatterSteps) {
```

Comment:
@nedtwigg / @jbduncan Here I'm not quite sure if I got the process of settting up a FormatterStep "pipeline" quite right. Is this the way to do it or did I mis-adapt the code from `gradle-plugin` and `maven-plugin`?

Reply:
This looks correct to me. Open the formatter as AutoCloseable and use LintState, those are the two main things and you did them both.